state-commitments/optimized/inclusion_proof.go

```diff
@@ -0,0 +1,113 @@
+	if len(leaves) == 0 {
+		return nil, nil
+	}
+	ok, err := isPowTwo(virtual)
+	if err != nil {
+		return nil, err
+	}
+	if !ok {
+		return nil, errors.New("virtual size must be a power of 2")
+	}
+	if leafIndex >= uint64(len(leaves)) {
+		return nil, errors.New("leaf index out of bounds")
+	}
+	if virtual < uint64(len(leaves)) {
+		return nil, errors.New("virtual size must be greater than or equal to the number of leaves")
+	}
+	numRealLeaves := uint64(len(leaves))
+	lastLeaf, err := h.hash(leaves[numRealLeaves-1][:])
+	if err != nil {
+		return nil, err
+	}
+	depth := int(math.Ceil(math.Log2(float64(virtual))))
+
+	// Precompute virtual hashes
+	virtualHashes, err := h.precomputeRepeatedHashes(&lastLeaf, depth)
+	if err != nil {
+		return nil, err
+	}
+	var proof []common.Hash
+	for level := 0; level < depth; level++ {
+		nodeIndex := leafIndex >> level
+		siblingHash, exists, err := h.computeSiblingHash(nodeIndex, uint64(level), numRealLeaves, virtual, leaves, virtualHashes)
+		if err != nil {
+			return nil, err
+		}
+		if exists {
+			proof = append(proof, siblingHash)
+		}
+	}
+	return proof, nil
+}
+
+// Computes the hash of a node's sibling at a given index and level.
+func (h *HistoryCommitter) computeSiblingHash(
+	nodeIndex uint64,
+	level uint64,
+	N uint64,
+	virtual uint64,
+	hLeaves []common.Hash,
+	hNHashes []common.Hash,
+) (common.Hash, bool, error) {
+	siblingIndex := nodeIndex ^ 1
+	// Essentially ceil(virtual / (2 ** level))
+	numNodes := (virtual + (1 << level) - 1) / (1 << level)
+	if siblingIndex >= numNodes {
+		// No sibling exists, so use a zero hash.
+		return common.Hash{}, false, nil
+	} else if siblingIndex >= paddingStartIndexAtLevel(N, level) {
+		return hNHashes[level], true, nil
+	} else {
+		siblingHash, err := h.computeNodeHash(siblingIndex, level, N, hLeaves, hNHashes)
+		if err != nil {
+			return emptyHash, false, err
+		}
+		return siblingHash, true, nil
+	}
+}
+
+// Recursively computes the hash of a node at a given index and level.
+func (h *HistoryCommitter) computeNodeHash(
+	nodeIndex uint64, level uint64, numRealLeaves uint64, leaves []common.Hash, virtualHashes []common.Hash,
+) (common.Hash, error) {
+	if level == 0 {
+		if nodeIndex >= numRealLeaves {
+			// Node is in padding (the virtual segment of the tree).
+			return virtualHashes[0], nil
+		} else {
+			return h.hash(leaves[nodeIndex][:])
+		}
+	} else {
+		if nodeIndex >= paddingStartIndexAtLevel(numRealLeaves, level) {
+			return virtualHashes[level], nil
+		} else {
+			leftChild, err := h.computeNodeHash(2*nodeIndex, level-1, numRealLeaves, leaves, virtualHashes)
+			if err != nil {
+				return emptyHash, err
+			}
+			rightChild, err := h.computeNodeHash(2*nodeIndex+1, level-1, numRealLeaves, leaves, virtualHashes)
+			if err != nil {
+				return emptyHash, err
+			}
+			data := append(leftChild.Bytes(), rightChild.Bytes()...)
```

Comment:
I think this is safe with the current implementation of `Bytes()`, but it's a bit scary, because if the capacity of the bytes returned by `Bytes` is at least 64 bytes then it'll end up mutating it.

Reply:
This is fixed with the new hash method that operates on a `...*common.Hash` which are known to be `[32]byte` instead of these `[]byte` which could (in theory, be longer.)